squad/core/models.py

```diff
@@ -203,25 +203,43 @@ def non_important_metadata(self):
     @property
     def finished(self):
         """
-        A finished build is a build that has at least N test runs for each of
-        the project environments, where N is configured in
-        Environment.expected_test_runs.
-
-        If an environment does not have expected_test_runs set, or has it set to
-        0, then there must be at least one test run for that environment.
+        A finished build is a build that satisfies one of the following conditions:
+
+        * it has no pending CI test jobs.
+        * it has no submitted CI test jobs, and has at least N test runs for each of
+          the project environments, where N is configured in
+          Environment.expected_test_runs.  the expected number of completed
+          test runs per environment. If an environment does not have
+          expected_test_runs set, or has it set to 0, then there must be at
+          least one test run for that environment.
         """
-        environments = self.project.environments.all()
-        expected = {e.id: e.expected_test_runs for e in environments}
+        # XXX note that by using test_jobs here, we are adding an implicit
+        # dependency on squad.ci, what in theory violates our architecture.
+        testjobs = self.test_jobs
+        if testjobs.count() > 0:
+            # If this build has CI jobs, then it's finished if there are no
+            # outstanding test jobs.
+            return testjobs.filter(fetched=False).count() == 0
```

Comment:
This might cause issues in case there are both CI and non-CI results reported to the same project. In this case all non-automated test results have to be reported between submitting the automated TestJob and completing it. I think we need at least a paragraph on that in the docs.